File: android/PositiveOnlySocial/app/src/main/java/com/example/positiveonlysocial/models/viewmodels/TagFeedViewModel.kt
package com.example.positiveonlysocial.models.viewmodels

import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.positiveonlysocial.api.PositiveOnlySocialAPI
import com.example.positiveonlysocial.data.model.Post
import com.example.positiveonlysocial.data.model.UserSession
import com.example.positiveonlysocial.data.security.KeychainHelperProtocol
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import kotlin.collections.emptyList

private const val TAG = "TagFeedViewModel"

/**
 * Drives the tag feed (issue #379): the posts carrying a given #hashtag,
 * paginated. Mirrors FeedViewModel, swapping the feed fetch for the
 * browse-by-tag endpoint.
 */
class TagFeedViewModel(
    private val api: PositiveOnlySocialAPI,
    private val keychainHelper: KeychainHelperProtocol,
    private val tag: String,
    private val account: String = "userSessionToken"
) : ViewModel() {

    private val _posts = MutableStateFlow<List<Post>>(emptyList())
    val posts: StateFlow<List<Post>> = _posts.asStateFlow()

    private val _isLoadingNextPage = MutableStateFlow(false)
    val isLoadingNextPage: StateFlow<Boolean> = _isLoadingNextPage.asStateFlow()

    private val _isRefreshing = MutableStateFlow(false)
    val isRefreshing: StateFlow<Boolean> = _isRefreshing.asStateFlow()

    /** Like / report / retract-report / delete for the posts in this tag feed,
     * so they can be acted on without opening each one (issue #267). */
    val postActions = PostListActions(api, keychainHelper, viewModelScope, _posts, account)

    private var canLoadMore = true
    private var currentPage = 0
    private val service = "positive-only-social.Positive-Only-Social"

    fun refresh() {
        if (_isRefreshing.value || _isLoadingNextPage.value) return

        _isRefreshing.value = true

        viewModelScope.launch {
            try {
                val userSession = keychainHelper.load(UserSession::class.java, service, account)
                if (userSession == null) {
                    Log.e(TAG, "No active session found — cannot refresh tag feed")
                    return@launch
                }

                val response = api.getPostsByTag(userSession.sessionToken, tag, 0)
                if (response.isSuccessful) {
                    val newPosts = response.body() ?: emptyList()
                    _posts.value = newPosts
                    canLoadMore = newPosts.isNotEmpty()
                    currentPage = if (newPosts.isEmpty()) 0 else 1
                } else {
                    Log.e(TAG, "Failed to refresh tag feed: ${response.errorBody()?.string()}")
                }
            } catch (e: Exception) {
                Log.e(TAG, "Failed to refresh tag feed", e)
            } finally {
                _isRefreshing.value = false
            }
        }
    }

    fun fetchNextPage() {
        if (_isLoadingNextPage.value || _isRefreshing.value || !canLoadMore) return

        _isLoadingNextPage.value = true

        viewModelScope.launch {
            try {
                val userSession = keychainHelper.load(UserSession::class.java, service, account)
                if (userSession == null) {
                    Log.e(TAG, "No active session found — cannot fetch tag feed")
                    return@launch
                }

                val response = api.getPostsByTag(userSession.sessionToken, tag, currentPage)
                if (response.isSuccessful) {
                    val newPosts = response.body() ?: emptyList()
                    if (newPosts.isEmpty()) {
                        canLoadMore = false
                    } else {
                        _posts.value += newPosts
                        currentPage += 1
                    }
                } else {
                    Log.e(TAG, "Failed to fetch tag feed: ${response.errorBody()?.string()}")
                }
            } catch (e: Exception) {
                Log.e(TAG, "Failed to fetch tag feed", e)
            } finally {
                _isLoadingNextPage.value = false
            }
        }
    }
}

File: ios/Positive Only Social/Positive Only Social/VibesViewModels/TagFeedViewModel.swift
//
//  TagFeedViewModel.swift
//  Positive Only Social
//
//  Drives the tag feed (issue #379): the posts carrying a given #hashtag,
//  paginated. Mirrors FeedViewModel, swapping the feed fetch for the
//  browse-by-tag endpoint.
//

import Foundation
import Combine

@MainActor
final class TagFeedViewModel: ObservableObject {
    private let api: Networking
    private let keychainHelper: KeychainHelperProtocol
    private let account: String
    private let keychainService = GVOAppConstants.keychainService
    let tag: String
    @Published var posts: [Post] = []
    @Published var isLoadingNextPage = false
    private var canLoadMore = true
    private var currentPage = 0

    // Drops a post from the list when it is deleted from its detail view, so
    // the tag feed reflects the removal without a full reload.
    private var postDeletedCancellable: AnyCancellable?

    convenience init(tag: String, api: Networking, keychainHelper: KeychainHelperProtocol) {
        self.init(tag: tag, api: api, keychainHelper: keychainHelper, account: "userSessionToken")
    }

    init(tag: String, api: Networking, keychainHelper: KeychainHelperProtocol, account: String,
         notificationCenter: NotificationCenter = .default) {
        self.tag = tag
        self.api = api
        self.keychainHelper = keychainHelper
        self.account = account

        postDeletedCancellable = notificationCenter.publisher(for: .postDeleted)
            .receive(on: RunLoop.main)
            .sink { [weak self] notification in
                guard let postIdentifier = notification.object as? String else { return }
                self?.posts.removeAll { $0.id == postIdentifier }
            }
    }

    func fetchNextPage() {
        guard !isLoadingNextPage && canLoadMore else { return }
        isLoadingNextPage = true
        Task { await loadPage(currentPage, replacingExisting: false) }
    }

    /// Reloads the tag feed from the first page. `async` so `.refreshable`
    /// keeps the spinner up until the fresh posts arrive; the cursor is only
    /// reset on a successful first page so a failed refresh can't duplicate it.
    func refresh() async {
        guard !isLoadingNextPage else { return }
        isLoadingNextPage = true
        await loadPage(0, replacingExisting: true)
    }

    private func loadPage(_ page: Int, replacingExisting: Bool) async {
        do {
            guard let userSession = try keychainHelper.load(UserSession.self, from: keychainService, account: account) else {
                NSLog("%@", "No active session found — cannot fetch tag feed")
                isLoadingNextPage = false
                return
            }

            let responseData = try await api.getPostsForTag(
                sessionManagementToken: userSession.sessionToken, tag: tag, batch: page)
            let newPosts = try JSONDecoder().decode([Post].self, from: responseData)

            if replacingExisting {
                posts = newPosts
                canLoadMore = !newPosts.isEmpty
                currentPage = newPosts.isEmpty ? 0 : 1
            } else if newPosts.isEmpty {
                canLoadMore = false
            } else {
                posts.append(contentsOf: newPosts)
                currentPage += 1
            }
        } catch {
            NSLog("%@", "Failed to fetch tag feed: \(error)")
        }
        isLoadingNextPage = false
    }
}
